internal/sync/cache.go

```diff
@@ -0,0 +1,362 @@
+)
+
+type ResourceInfo struct {
+	GitSyncName string
+
+	Health       *health.HealthStatus
+	manifestHash string
+}
+
+type LiveStateCache interface {
+	// GetManagedLiveObjs Returns state of live nodes which correspond for target nodes of specified gitsync.
+	GetManagedLiveObjs(gitsync *v1alpha1.GitSync, targetObjs []*unstructured.Unstructured) (map[kube.ResourceKey]*unstructured.Unstructured, error)
+	// Init must be executed before cache can be used
+	Init() error
+}
+
+type cacheSettings struct {
+	clusterSettings clustercache.Settings
+
+	// ignoreResourceUpdates is a flag to enable resource-ignore rules.
+	ignoreResourceUpdatesEnabled bool
+}
+
+type ObjectUpdatedHandler = func(managedByGitSync map[string]bool, ref v1.ObjectReference)
+
+type liveStateCache struct {
+	clusterCacheConfig *rest.Config
+	logger             *zap.SugaredLogger
+	onObjectUpdated    ObjectUpdatedHandler
+
+	cluster       clustercache.ClusterCache
+	cacheSettings cacheSettings
+	lock          sync.RWMutex
+}
+
+func NewLiveStateCache(
+	onObjectUpdated ObjectUpdatedHandler,
+) LiveStateCache {
+	return &liveStateCache{
+		onObjectUpdated: onObjectUpdated,
+	}
+}
+
+func (c *liveStateCache) getCluster() (clustercache.ClusterCache, error) {
+	c.lock.RLock()
+	cacheSettings := c.cacheSettings
+	c.lock.RUnlock()
+
+	if c.cluster != nil {
+		return c.cluster, nil
+	}
+
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	if c.cluster != nil {
+		return c.cluster, nil
+	}
+
+	clusterCacheConfig := c.clusterCacheConfig
+
+	// TODO: check if we need set managed namespace,
+	//       also enable logging within clusterCache
+	clusterCacheOpts := []clustercache.UpdateSettingsFunc{
+		clustercache.SetListSemaphore(semaphore.NewWeighted(clusterCacheListSemaphoreSize)),
+		clustercache.SetListPageSize(clusterCacheListPageSize),
+		clustercache.SetListPageBufferSize(clusterCacheListPageBufferSize),
+		clustercache.SetWatchResyncTimeout(clusterCacheWatchResyncDuration),
+		clustercache.SetClusterSyncRetryTimeout(clusterSyncRetryTimeoutDuration),
+		clustercache.SetResyncTimeout(clusterCacheResyncDuration),
+		clustercache.SetSettings(cacheSettings.clusterSettings),
+		clustercache.SetPopulateResourceInfoHandler(c.PopulateResourceInfo),
+		clustercache.SetRetryOptions(clusterCacheAttemptLimit, clusterCacheRetryUseBackoff, isRetryableError),
+	}
+
+	clusterCache := clustercache.NewClusterCache(clusterCacheConfig, clusterCacheOpts...)
+
+	_ = clusterCache.OnResourceUpdated(func(newRes *clustercache.Resource, oldRes *clustercache.Resource, namespaceResources map[kube.ResourceKey]*clustercache.Resource) {
+		c.lock.RLock()
+		cacheSettings := c.cacheSettings
+		c.lock.RUnlock()
+
+		if cacheSettings.ignoreResourceUpdatesEnabled && oldRes != nil && newRes != nil && skipResourceUpdate(resInfo(oldRes), resInfo(newRes)) {
+			return
+		}
+
+		// TODO: when switch to change event triggering for sync tasking, notify
+		//       the GitSync managing the resources that there are changes happen
```

Comment:
I guess you will call `c.onObjectUpdated()` here?

Reply:
yes